design-docs/cameraserver.adoc

```diff
@@ -0,0 +1,419 @@
+cameras).  URL should be of the form
+`"http://user:pass@server:port/path/to/stream"`.
+
+==== CreateUserInput
+
+[source,c]
+CS_Input CS_CreateUserInput(int nChannels);
+
+==== DestroyInput
+
+[source,c]
+void CS_DestroyInput(CS_Input input);
+
+==== EnumerateInputs
+
+==== AddInputListener
+
+[source,c]
+CS_InputListener CS_AddInputListener(void (*callback) (CS_Input input, const CS_InputInfo* info, int event), int eventMask);
+
+Notifies a callback function when an input is created, destroyed, or
+changes state.  The eventMask is a bitmask which specifies what events
+should cause a callback to occur.
+
+[cols="1,3"]
+|===
+|Event |Description
+
+|Create
+|New input created
+
+|Destroy
+|Input destroyed
+
+|Connected
+|Input connected (e.g. camera successfully connected)
+
+|Disconnected
+|Input disconnected (lost IP connection, USB camera disconnected)
+|===
+
+
+==== RemoveInputListener
+
+[source,c]
+void CS_RemoveInputListener(CS_InputListener listener);
+
+=== Output Management
+
+==== CreateHTTPOutput
+
+[source,c]
+CS_Output CS_CreateHTTPOutput(const char* listenAddress, int port);
+
+Creates a M-JPEG over HTTP server at the specified port.  The server
+allows multiple clients to connect to the port (this will effectively
+act as multiple virtual outputs listening to the same input that is
+configured for this output).
+
+==== CreateRTSPOutput
+
+[source,c]
+CS_Output CS_CreateRTSPOutput(const char* listenAddress, int port);
+
+Creates a M-JPEG over RTP server at the specified port.  The RTSP
+server will be at the specified port, but data will be sent via M-JPEG
+over UDP in accordance with <<rfc2435>>.
+
+==== DestroyOutput
+
+[source,c]
+void CS_DestroyOutput(CS_Output output);
+
+==== EnumerateOutputs
+
+==== AddOutputListener
+
+==== RemoveOutputListener
+
+=== User Input Functions
+
+These functions are only valid for use with inputs created with
+CreateUserInput().
+
+==== PutImage
+
+[source,c]
+CS_PutImage(CS_Input input, int channel, CvMat* image);
+
+Puts an OpenCV image as the latest image for an user input and
+channel.  This does *not* notify outputs that a new image is
+available; NotifyFrame() must be called to do that.
+
+==== NotifyFrame
+
+[source,c]
+void CS_NotifyFrame(CS_Input input);
+
+Notifies outputs that a user input has a new frame available.
+
+=== Input Functions
+
+==== WaitForFrame
+
+[source,c]
+long CS_WaitForFrame(CS_Input input);  // returns frame timestamp
+
+Polled interface to wait for a new frame from an input (usually a
+camera).
+
+This function is blocking and does not return until a new frame has
+been received or the input has been destroyed by another thread or
+due to program shutdown (in which case 0 is returned).  In particular,
+this function does *not* return simply due to a camera disconnect (as
+this may be a temporary condition).
+
+==== GetImage
+
+[source,c]
+CS_ErrorCode CS_GetImage(CS_Input input, int channel, CvMat* image, long* timestamp);
```

Comment:
Good idea, as that's definitely the common case.  I may do it as a higher level API though rather than at this layer.